### PR TITLE
[`PEFT`] Adapt example scripts to use PEFT

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes: https://github.com/huggingface/diffusers/issues/5341 

Wandb logs: https://wandb.ai/younesbelkada/text2image-fine-tune/runs/tj5k8xtm?workspace=user-younesbelkada 

I have fine-tuned it on Pokemon blip captions and the model trained after 1200 steps can be found here: https://huggingface.co/ybelkada/sd-1.5-pokemon-lora-peft 

cc @sayakpaul @patrickvonplaten 

TODO:

- Update the SDXL LoRA script + add peft dependency on example tests workflow

### CODE REVIEW

File: examples/text_to_image/train_text_to_image_lora.py
Comment (younesbelkada) [inline examples/text_to_image/train_text_to_image_lora.py:442]:
```suggestion
    # Freeze the unet parameters before adding adapters
    for param in unet.parameters():
```

File: examples/text_to_image/train_text_to_image_lora.py
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_lora.py:38]: Users might not be aware of this dependency. Should we do `is_peft_available()` here? And add `peft` to `requirements.txt`?

File: examples/text_to_image/train_text_to_image_lora.py
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_lora.py:447]: Target modules should be either fixed or be made CLI arg IMO. 

For our training scripts with LoRA, we only focus on the attention matrices (key, query, and value).

**Comment by sayakpaul** [comment]
Left some comments. Thanks for initiating this!

File: examples/text_to_image/train_text_to_image_lora.py
Comment (younesbelkada) [reply examples/text_to_image/train_text_to_image_lora.py:38]: Ah makes sense, will do!

**Comment by patrickvonplaten** [comment]
Before doing this change it would be good to have both a PEFT and a Transformers release to not force the user to install these packages from main

**Comment by younesbelkada** [comment]
> Before doing this change it would be good to have both a PEFT and a Transformers release to not force the user to install these packages from main

OK I will push the changes for SDXL and wait until the next release of PEFT before merging this PR

**Comment by younesbelkada** [comment]
I can confirm the scripts work fine under peft + diffusers main! I'll mark the PR as ready for review once do a release on PEFT

**Comment by younesbelkada** [comment]
This PR is ready for review! Let me know if you want me to work on saving adapter configs here, otherwise i can do it as a follow up PR. Please also see the attached wandb logs in the PR description above

File: examples/dreambooth/requirements.txt
Comment (sayakpaul) [inline examples/dreambooth/requirements.txt:7]: Should we specify a minimum version here?

File: examples/dreambooth/train_dreambooth_lora.py
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:820]: This can be deleted too!

File: examples/dreambooth/train_dreambooth_lora.py
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:826]: Is this comment still relevant?

File: examples/dreambooth/train_dreambooth_lora.py
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:904]: Very nice!

File: examples/dreambooth/train_dreambooth_lora.py
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:1282]: Let's maintain uniform naming conventions here. Since `unet_lora_state_dict` is more appropriate, let's go with `text_encoder_lora_state_dict` too?

Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:1282]: Also, won't we need to copy this logic over to the saving and loading hooks for intermediate checkpointing?

**Review by sayakpaul — COMMENTED** [review]
Wow! You made it look simpler than what I'd imagine!

For me, this PR is already in a good shape. Main points to address:

* Modify the READMEs to state about this update. 
* Add `peft` to the requirements for `text_to_image`. Also, note that if we separate requirements' files for SD and SDXL. 
* Uniformity in the variable naming (`unet_lora_state_dict`, `text_encoder_lora_state_dict`, etc.). 
* ~Go through the saving and loading hooks to make sure they have parity with how we're serializing the final state dicts here.~

Also, can we make sure if the following works even when `peft` is not installed?

```python
from diffusers import StableDiffusionXLPipeline
import torch

pipeline = StableDiffusionXLPipeline.from_pretrained(...)
pipeline.load_lora_weights("...") # load the final state dict
```

@patrickvonplaten do take a look here too. For now, we're serializing in the `diffusers` native format. But can also output the `peft` state dict and the adapter config with the `diffusers` state dict too.

**Comment by younesbelkada** [comment]
Thanks very much for your review and comments @sayakpaul - OK will address everything asap !

File: examples/dreambooth/requirements.txt
Comment (younesbelkada) [reply examples/dreambooth/requirements.txt]: Yeah definitely, just added it!

File: examples/dreambooth/train_dreambooth_lora.py
Comment (younesbelkada) [reply examples/dreambooth/train_dreambooth_lora.py:826]: Nope! removed it

File: examples/dreambooth/train_dreambooth_lora.py
Comment (younesbelkada) [reply examples/dreambooth/train_dreambooth_lora.py]: Makes sense! Done!

**Comment by younesbelkada** [comment]
@sayakpaul this PR is ready now, I can confirm loading the lora weights after training works as expected, I also pushed the weights on the hub so that you can try out:

```python
from diffusers import DiffusionPipeline
import torch

pipe = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16)
pipe.to("cuda")

pipe.load_lora_weights("ybelkada/test-lora-pokemon") # load the final state dict
```

Regarding saving adapter configs, I think we should think about it carefully to properly deal with the case we attach lora to text encoder as well or the multi-adapter case. I propose to address that in a follow up PR!

**Comment by sayakpaul** [comment]
@younesbelkada can you also check if training the text encoder works as expected too?

**Comment by younesbelkada** [comment]
Hi @sayakpaul I have confirmed it works also when trainign the text encoder and I pushed the trained weights on the Hub
```python
from diffusers import StableDiffusionXLPipeline
import torch

pipe = StableDiffusionXLPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16)
pipe.to("cuda")

pipe.load_lora_weights("ybelkada/test-pokemon-lora-sdxl-with-text-encoder") # load the final state dict
```
Note saving / loading is already tested directly in the CPU examples tests so we should not worry about it

> Could you elaborate a bit here? The text encoder LoRA should be attached in the similar manner no?

Yes, however you would have two separate `LoraConfig`, one for `text_encoder` and another for `unet` which may have different configurations. One thing we could do is to save those configs into separate subfolders. Let's discuss that on a dedicated thread on Slack with other PEFT maintainers and @patrickvonplaten

**Comment by sayakpaul** [comment]
Oh yes, makes sense. I had a discussion with @pacman100 about it months ago as well.

File: examples/dreambooth/requirements.txt
Comment (sayakpaul) [inline examples/dreambooth/requirements.txt:7]: Let's also add to https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/requirements_sdxl.txt. Applies for text-to-image, too.

**Review by sayakpaul — APPROVED** [review]
Looking very nice ❤️ Just one comment regarding adding `peft` as a dependency to `requirements_sdxl.text` too. 

@patrickvonplaten, in this PR, we're refactoring our existing LoRA training scripts to use `peft`. They feel cleaner and leaner to me. Also, most importantly, we're still serializing in the `diffusers` native format. I think we should revisit any serialization-related stuff in a later PR.

**Comment by sayakpaul** [comment]
The test failing is known and is being internally discussed.

**Comment by sayakpaul** [comment]
@patrickvonplaten could you please give this a review? Very much needed to prio this a bit. Also, take note of https://github.com/huggingface/diffusers/pull/5388#pullrequestreview-1735910253.

**Comment by BenjaminBossan** [comment]
> As discuss offline with @younesbelkada, I'm happy to merge this script as soon as the PEFT way of initialization matches what has been done in the legacy PEFT - the training should be just as easy as before. If this is the case, I'm happy to merge

Just to be clear, you mean the initialization of the LoRA parameters should be the same as in the non-PEFT diffusers LoRA implementation, instead of using the standard PEFT initialization, right? For now, this is not an available option in PEFT. I could look into adding it, but that means that it can only be used with a from source PEFT install or until we release the next PEFT version.

**Comment by sayakpaul** [comment]
> I could look into adding it, but that means that it can only be used with a from source PEFT install or until we release the next PEFT version.

I am okay with a source install (maybe even tagging the SHA commit).

**Comment by BenjaminBossan** [comment]
I created a PR on PEFT side: https://github.com/huggingface/peft/pull/1189

**Comment by sayakpaul** [comment]
@patrickvonplaten this should be more or less ready to be merged, especially with https://github.com/huggingface/peft/pull/1189. 

@BenjaminBossan if would like to review the `peft` related changes introduced in this PR, that would be awesome. 

Here's my Colab Notebook: https://colab.research.google.com/gist/sayakpaul/e112cc40288c5f0563efa4034dd50b72/scratchpad.ipynb.

File: examples/dreambooth/train_dreambooth_lora.py
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:864]: @younesbelkada in your previous iterations, we were missing "to_out.0" in the `target_module` for some scripts. I have added that.

**Comment by BenjaminBossan** [comment]
From the PEFT-side of things, this looks pretty good, thanks Sayak.

Just a consideration: Right now, the requirements include a source install of PEFT. This works for the moment but could be dangerous in the future, in case we add something to PEFT `main` which breaks the examples (of course, that wouldn't be intentional, but these things can happen). So I wonder if it would be more cautious to fix the hash of the PEFT install and add a comment that we should change the dependency to a proper PEFT version once the next PEFT release is out (most likely, that would be `>=0.7.0`).

**Comment by sayakpaul** [comment]
> So I wonder if it would be more cautious to fix the hash of the PEFT install and add a comment that we should change the dependency to a proper PEFT version once the next PEFT release is out (most likely, that would be >=0.7.0).

@BenjaminBossan good point. Could you get me the required hash here? I will resolve ASAP.

**Comment by BenjaminBossan** [comment]
> Could you get me the required hash here? I will resolve ASAP.

Latest hash is `8298f1a3668604ac9bc3f6e28b24e8eb554891a1`, so adding `@` + the hash to the requirement should do the trick. This is of course not required if we do:

> I'd like to wait a bit until PEFT has done a release to make it easier for users

**Comment by sayakpaul** [comment]
> Does the training now match more or less (e.g. is the new init used?).

@patrickvonplaten yes it does. Have verified it. This PR uses the new init method from peft too. I don't mind waiting either.

**Comment by younesbelkada** [comment]
PEFT 0.7.0 has been released ! https://github.com/huggingface/peft/releases/tag/v0.7.0
I have updated the requirements file to reflect the correct PEFT version - requesting a final review cc @patrickvonplaten @sayakpaul

**Review by patrickvonplaten — COMMENTED** [review]
Ok for me to add once @sayakpaul gives the green light

**Pull request merged**
